### PR TITLE
first take on dimensions API

### PR DESCRIPTION
**Pull request by UpcraftLP**
first work on a dimensions API that allows for easy travel between dimensions.

## additions:
- a pre teleportation event
- mods have the ability to register custom entity placers for their dimensions

## TODO / bugs:
- [x] ~loading existing worlds crashes the game~ FIXED!
- [ ] post teleportation event?

### CODE REVIEW

**Comment by asiekierka** [comment]
The patch is pretty heavy - and IMO confusing.

**Comment by UpcraftLP** [comment]
> The patch is pretty heavy - and IMO confusing.

Confusing? in what way?

**Comment by asiekierka** [comment]
It uses a lot of method_ names without really explaining why, same with why it copies a few dozen lines of vanilla logic that could always change

**Comment by UpcraftLP** [comment]
I'd map those names, but I know the forge mappings in that area too much..

**Comment by UpcraftLP** [comment]
Update: fixed the magic out of memory crash (idk how, but it's fixed)

File: src/main/java/net/fabricmc/fabric/mixin/dimension/MixinEntity.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;
+import org.spongepowered.asm.mixin.injection.callback.LocalCapture;
+
+@Mixin(Entity.class)
+public abstract class MixinEntity {
+
+    @Inject(method = "changeDimension", at = @At(value = "INVOKE_STRING", target = "Lnet/minecraft/util/profiler/Profiler;push(Ljava/lang/String;)V", args = "ldc=reposition", shift = At.Shift.AFTER), cancellable = true, locals = LocalCapture.CAPTURE_FAILHARD)
+    private void changeDimension(DimensionType targetType, CallbackInfoReturnable<Entity> cir, MinecraftServer server, DimensionType fromType, ServerWorld fromWorld, ServerWorld toWorld) {
```
Comment (modmuss50) [inline src/main/java/net/fabricmc/fabric/mixin/dimension/MixinEntity.java]: Is all this really needed! This is too much "magic" code imo.

File: src/main/java/net/fabricmc/fabric/impl/accessors/PlayerDimensionChangeAcessor.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.impl.accessors;
+
+public interface PlayerDimensionChangeAcessor {
```
Comment (Juuxel) [inline src/main/java/net/fabricmc/fabric/impl/accessors/PlayerDimensionChangeAcessor.java]: Typo here?

File: src/main/java/net/fabricmc/fabric/api/event/dimension/AttemptDimensionTeleportCallback.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+            }
+        }
+        return currentResult;
+    });
+
+    /**
+     * @return <ul>
+     *     <li>{@link EventResult#SKIP_FURTHER_PROCESSING} if you don't want to have the default {@link EntityPlacer} handle entity placement and portal creation</li>
+     *     <li>{@link EventResult#CANCEL_TELEPORT} if you want to prevent the entity from teleporting at all</li>
+     *     <li>{@link EventResult#DEFAULT} if you want the default {@link EntityPlacer} handle placement of the entity</li>
```
Comment (Pyrofab) [inline src/main/java/net/fabricmc/fabric/api/event/dimension/AttemptDimensionTeleportCallback.java]: Wouldn't this javadoc be more adequate on the enum itself (with a `@see` link) ? Javadoc on the method should rather explain the behaviour of the method itself (when it is called, what it should do...)

File: src/main/java/net/fabricmc/fabric/mixin/dimension/MixinEntity.java (hunk 1 above)
Comment (UpcraftLP) [reply src/main/java/net/fabricmc/fabric/mixin/dimension/MixinEntity.java]: wdym "magic"?
thing is, vanilla already has `Entity#changeDImension()`, and uses it to transfer entities between worlds.
Because of that it wouldn#t make sense to have a static `FabricDimensionHelper.transferEntityTo()` method of sorts, because that#d mean modders would have to implement dimension changes TWICE. once for vanilla, once for fabric.

File: src/main/java/net/fabricmc/fabric/impl/accessors/PlayerDimensionChangeAcessor.java (hunk 2 above)
Comment (UpcraftLP) [reply src/main/java/net/fabricmc/fabric/impl/accessors/PlayerDimensionChangeAcessor.java]: yes OwO

File: src/main/java/net/fabricmc/fabric/api/event/dimension/AttemptDimensionTeleportCallback.java (hunk 3 above)
Comment (UpcraftLP) [reply src/main/java/net/fabricmc/fabric/api/event/dimension/AttemptDimensionTeleportCallback.java]: good point, will fix.

File: src/main/java/net/fabricmc/fabric/impl/accessors/PlayerDimensionChangeAcessor.java (hunk 2 above)
Comment (UpcraftLP) [reply src/main/java/net/fabricmc/fabric/impl/accessors/PlayerDimensionChangeAcessor.java]: it#s a leftover from earlier attempty anyways, fixed it now.

File: src/main/java/net/fabricmc/fabric/api/event/dimension/AttemptDimensionTeleportCallback.java (hunk 3 above)
Comment (UpcraftLP) [reply src/main/java/net/fabricmc/fabric/api/event/dimension/AttemptDimensionTeleportCallback.java]: fixed.

**Comment by Boundarybreaker** [comment]
Needs another look

**Comment by UpcraftLP** [comment]
@Boundarybreaker fixed

**Comment by UpcraftLP** [comment]
fixed again.

File: src/main/java/net/fabricmc/fabric/api/event/dimension/AttemptDimensionTeleportCallback.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+    /**
+     * fired to allow mods to intercept the teleportation of an entity to another dimension
+     * this hook can take priority over default behaviour, or even cancel the teleportation
+     *
+     * @return an {@link EventResult} to determine further processing of the teleport
+     */
+    EventResult placeEntity(Entity entity, DimensionType from, DimensionType to);
+
+    enum EventResult {
```
Comment (Boundarybreaker) [inline src/main/java/net/fabricmc/fabric/api/event/dimension/AttemptDimensionTeleportCallback.java]: I think these should be called `CONTINUE`, `CANCEL`, and `DEFAULT`, probably. Modders should already know "continue" as "skip to the next step", and `CANCEL_TELEPORT` is a bit more verbose than needed.

File: src/main/java/net/fabricmc/fabric/mixin/client/texture/MixinSpriteAtlasTexture.java [hunk 5]
```diff
@@ -40,14 +41,12 @@
 
 @Mixin(SpriteAtlasTexture.class)
 public abstract class MixinSpriteAtlasTexture {
```
Comment (Boundarybreaker) [inline src/main/java/net/fabricmc/fabric/mixin/client/texture/MixinSpriteAtlasTexture.java]: Any reason these changes are needed?

File: src/main/java/net/fabricmc/fabric/mixin/dimension/MixinEntity.java (hunk 1 above)
Comment (Boundarybreaker) [reply src/main/java/net/fabricmc/fabric/mixin/dimension/MixinEntity.java]: I think this is necessary enough to keep.

File: src/main/java/net/fabricmc/fabric/mixin/resources/MixinMinecraftGame.java [hunk 6]
```diff
@@ -16,64 +16,54 @@
...
 import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
 import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;
 import org.spongepowered.asm.mixin.injection.callback.LocalCapture;
 
-import java.util.ArrayList;
 import java.util.List;
 import java.util.concurrent.CompletableFuture;
 
 @Mixin(MinecraftClient.class)
 public class MixinMinecraftGame {
```
Comment (Boundarybreaker) [inline src/main/java/net/fabricmc/fabric/mixin/resources/MixinMinecraftGame.java]: These changes seem a bit out of scope for the current PR. While I'm not opposed to them, I think it'd better to move them out for the time being.

File: src/main/java/net/fabricmc/fabric/api/event/dimension/AttemptDimensionTeleportCallback.java (hunk 4 above)
Comment (UpcraftLP) [reply src/main/java/net/fabricmc/fabric/api/event/dimension/AttemptDimensionTeleportCallback.java]: done

**Comment by Boundarybreaker** [comment]
How much more work does this need?

**Comment by modmuss50** [comment]
Im still not convinced all of this is needed, In Simple Void world I was able to make it work with one very simple mixin, 

https://github.com/modmuss50/SimpleVoidWorld/blob/1.14/src/main/java/me/modmuss50/svw/mixin/MixinPortalForcer.java

**Comment by asiekierka** [comment]
Should something akin to this mixin end up in Fabric, then?

**Comment by modmuss50** [comment]
Most likey yes, im not sure if this works with none player entities however, that can be tested.

**Comment by valoeghese** [comment]
:crab: Pr #110 is GONE :crab:

**Comment by grondag** [comment]
Is this still a going concern in light of #256?

**Comment by Draylar** [comment]
After talking with Upcraft and Ansraer, I decided to re-work this PR into a new one, which should be out later today for an initial review. It's fine to close this one.
